File: WhiteBoardCast/src/main/java/com/livejournal/karino2/multigallery/CacheEngine.java
package com.livejournal.karino2.multigallery;

import android.graphics.Bitmap;

import java.util.ArrayList;
import java.util.HashMap;

/**
* Created by karino on 11/29/13.
*/
public class CacheEngine {
    ArrayList<String> keys = new ArrayList<String>();
    HashMap<String, Bitmap> entries = new HashMap<String, Bitmap>();

    int cacheSize = 100; // default size

    CacheEngine() {
    }

    public CacheEngine(int size) {
        cacheSize = size;
    }

    Bitmap lookupByMediaItem(MediaItem item) {
        return lookup(item.getPath());
    }

    public Bitmap lookup(String path) {
        if(entries.containsKey(path))
            return entries.get(path);
        return null;
    }

    void put(MediaItem item, Bitmap thumbnail) {
        put(item.getPath(), thumbnail);
    }

    public void put(String path, Bitmap thumbnail) {
        if(keys.contains(path)) {
            keys.remove(path);
            keys.add(path);
            return;
        }

        keys.add(path);
        entries.put(path, thumbnail);

        if(keys.size() > cacheSize) {
            String first = keys.get(0);
            entries.remove(first);
            keys.remove(0);
        }
    }
}

File: WhiteBoardCast/src/main/java/com/livejournal/karino2/whiteboardcast/CacheEngine.kt
package com.livejournal.karino2.whiteboardcast

import android.graphics.Bitmap


/**
 * Created by karino on 11/29/13.
 */
class CacheEngine {
    var keys = ArrayList<String>()
    var entries = HashMap<String, Bitmap>()
    var cacheSize = 100 // default size

    internal constructor()
    constructor(size: Int) {
        cacheSize = size
    }

    fun lookup(path: String): Bitmap? {
        return if (entries.containsKey(path)) entries[path] else null
    }


    fun put(path: String, thumbnail: Bitmap) {
        if (keys.contains(path)) {
            keys.remove(path)
            keys.add(path)
            return
        }
        keys.add(path)
        entries[path] = thumbnail
        if (keys.size > cacheSize) {
            val first = keys[0]
            entries.remove(first)
            keys.removeAt(0)
        }
    }
}
